### PR TITLE
Use pymatgen for pbc neighbor search speed up

### PR DESCRIPTION
**Pull request by corochann**

### CODE REVIEW

File: torch_dftd/functions/edge_extraction.py [hunk 1]
```diff
@@ -46,14 +110,10 @@ def calc_edge_index(
...
-            cell.detach().cpu().numpy(),
-            pos.detach().cpu().numpy(),
-            cutoff,
-        )
-        edge_index = torch.tensor(np.stack([idx_i, idx_j], axis=0), device=pos.device)
-        # convert int64 -> pos.dtype (float)
-        S = torch.tensor(S, dtype=pos.dtype, device=pos.device)
+        try:
+            edge_index, S = calc_neighbor_by_pymatgen(pos, cell, pbc, cutoff)
+        except Exception:
```
Comment (So-Takamoto) [inline torch_dftd/functions/edge_extraction.py:115]: Can we limit the type of `Exception` here, e.g. `NotImplementedError`?

**Comment by So-Takamoto** [comment]
It may be good to add a test so that `calc_neighbor_by_ase` and `calc_neighbor_by_pymatgen` are equivalent.

File: torch_dftd/functions/edge_extraction.py [hunk 2]
```diff
@@ -4,9 +4,73 @@
...
+    edge_index = torch.tensor(
+        np.stack([c_index, n_index], axis=0), dtype=torch.long, device=pos.device
+    )
+    edge_distances = torch.tensor(n_distance, dtype=torch.float32, device=pos.device)
+    cell_offsets = torch.tensor(offsets, dtype=torch.long, device=pos.device)
+
+    # remove distances smaller than a tolerance ~ 0. The small tolerance is
+    # needed to correct for pymatgen's neighbor_list returning self atoms
+    # in a few edge cases.
+    nonzero = torch.where(edge_distances >= 1e-8)[0]
```
Comment (So-Takamoto) [inline torch_dftd/functions/edge_extraction.py:68]: Maybe it is good to enforce that `dtype` of `positions` is always `float64` because of this line.

File: torch_dftd/functions/edge_extraction.py (hunk 1 above)
Comment (corochann) [reply torch_dftd/functions/edge_extraction.py:115]: I wanted to consider the other case than `NotImplementedError`, but I think it's nice to raise exit with error then. I will change it.

File: torch_dftd/functions/edge_extraction.py (hunk 2 above)
Comment (So-Takamoto) [reply torch_dftd/functions/edge_extraction.py:68]: I am not sure where this small tolerance will come from, but it seems some kind of subtraction seems to occur in somewhere, which will reduce the numerical accuracy. 
For example, the difference between 1.0 and next to 1.0 in fp32 is log_10{2^-24}, which is around 6.0e-8 in decimal value. So I feel a bit dangerous about this value.

**Comment by corochann** [comment]
added test and simpllified impl.

File: tests/functions_tests/test_edge_extraction.py [hunk 3]
```diff
@@ -0,0 +1,49 @@
+import pytest
+import torch
+from torch_dftd.functions.edge_extraction import calc_neighbor_by_ase, calc_neighbor_by_pymatgen
+
+
+def test_calc_neighbor_equivalent():
+    n_nodes = 5
+    pbc = torch.tensor([True, True, True])
+    cell = torch.randn((3, 3))
```
Comment (So-Takamoto) [inline tests/functions_tests/test_edge_extraction.py:9]: It seems `cell` could be negative volume. I think that is OK. However, is it also possible that the volume is almost zero, for example, one of the lattice vector is almost zero or the rank of `cell` is less than 3?
In my understanding, the number of neighbors could be arbitrary large in those cases.

**Review by So-Takamoto — APPROVED** [review]
It looks good to me! I put one minor comment within the test.

**Comment by corochann** [comment]
Thank you, I will put issue about how to make the test more robust.
